File: backend/app/core/calibration.py
"""Calibration card — reasons this might be wrong (F16).

Derives genuine uncertainty sources and returns them as NoteItems.
Contributes 0 points and never changes the verdict.
"""

from __future__ import annotations

from app.models.schemas import CategoryResult, NoteItem, ParsedApk


def build_calibration_notes(
    apk: ParsedApk,
    category: CategoryResult,
    not_checked_checks: list[str] | None = None,
) -> list[NoteItem]:
    notes: list[NoteItem] = []

    if category.id == "unknown":
        notes.append(NoteItem(reason_key="calibration.unknown_category", params={}))
    elif category.confidence == "low":
        notes.append(NoteItem(reason_key="calibration.low_confidence", params={}))

    if not apk.has_launcher_activity:
        notes.append(NoteItem(reason_key="calibration.no_launcher", params={}))

    if not_checked_checks:
        notes.append(
            NoteItem(
                reason_key="calibration.not_checked_checks",
                params={"checks": ", ".join(not_checked_checks)},
            )
        )

    # General honesty caveat regarding allowlists / static analysis
    notes.append(NoteItem(reason_key="calibration.allowlist_caveat", params={}))

    return notes

File: shared/src/commonMain/kotlin/com/appautopsy/analysis/core/Calibration.kt
package com.appautopsy.analysis.core

import com.appautopsy.analysis.model.CategoryResult
import com.appautopsy.analysis.model.Confidence
import com.appautopsy.analysis.model.NoteItem
import com.appautopsy.analysis.model.ParsedApk

/**
 * Calibration card (F16): the reasons this verdict might be wrong.
 *
 * Calibrated honesty is rare in security tools and this is where it is
 * enforced. Every note is a genuine limitation of the analysis, derived from
 * what actually happened during the scan — never boilerplate hedging.
 *
 * Contributes 0 points and never changes the verdict.
 */
object Calibration {

    /**
     * @param notCheckedChecks names of checks that did not run, e.g.
     *   `reputation`. Naming them is required: a check that silently vanished
     *   looks identical to a check that passed, and that is a lie by omission.
     */
    fun notes(
        apk: ParsedApk,
        category: CategoryResult,
        notCheckedChecks: List<String> = emptyList(),
    ): List<NoteItem> {
        val notes = mutableListOf<NoteItem>()

        if (category.id == "unknown") {
            notes += NoteItem(reasonKey = "calibration.unknown_category")
        } else if (category.confidence == Confidence.LOW) {
            notes += NoteItem(reasonKey = "calibration.low_confidence")
        }

        if (!apk.hasLauncherActivity) {
            notes += NoteItem(reasonKey = "calibration.no_launcher")
        }

        if (notCheckedChecks.isNotEmpty()) {
            notes += NoteItem(
                reasonKey = "calibration.not_checked_checks",
                params = mapOf("checks" to notCheckedChecks.joinToString(", ")),
            )
        }

        // Always: an expected-permission allowlist is a judgement call, and
        // saying so is cheaper than being caught not saying it.
        notes += NoteItem(reasonKey = "calibration.allowlist_caveat")

        return notes
    }
}
